config/config.txt
# Logging (1=debug, 2=info, 3=warning, 4=error)
log_level=2

# Paths
log_path=./logs/castor.log
socket_path=./socket/castor.socket
audio_source_path=./audio/source
audio_playlist_path=./audio/playlist
audio_record_path=./audio/record
audio_fallback_path=./audio/fallback

# Caching time (sec.)
preload_time_file=3600
preload_time_fallback=3600

# Fallback Track Shuffling (random ordering each reload; 0 = off)
fallback_shuffle=1

# Fallback Test Signal (synthesized sine waves; 0 = off)
fallback_sine_synth=1

# Silence Detector
silence_threshold=-80
silence_start_duration=10
silence_stop_duration=1

# Fading
program_fade_in_time=1.0
program_fade_out_time=1.0
fallback_cross_fade_time=5.0

# Audio Device
in_device_name=Soundcraft Signature 12 MTK: USB Audio
out_device_name=Soundcraft Signature 12 MTK: USB Audio
sample_rate=44100

# Schedule Recording (leave path empty to disable)
# record_schedule_path=schedule
record_schedule_format=mp3
record_schedule_bitrate=320000

# Block Recording (leave path empty to disable)
# record_block_path=block
record_block_format=ogg
record_block_bitrate=128000
record_block_duration=3600

# Stream Output (leave stream_out_url empty to disable)
# stream_out_url=icecast://source:hackme@localhost:8000/radio.mp3
stream_out_metadata_url=http://source:hackme@localhost:8000/admin/metadata?mount=/radio.mp3
stream_out_name=Radio Station
stream_out_description=Live Stream
stream_out_genre=Electronic
stream_out_href=http://localhost:8000/radio.mp3

# API (comment-out to disable)
# program_url=http://aura.local/steering/api/v1/program/playout/
# media_url=http://aura.local/steering/api/v1/media/
# playlog_url=http://localhost:8010/api/v1/playlog
# health_url=http://localhost:8010/api/v1/source/health/1
# clock_url=http://localhost:8010/api/v1/clock
calendar_refresh_interval=60
calendar_cache_path=./cache/calendar.json
health_report_interval=10

# YARM MySQL API
# yarm_host=
# yarm_user=
# yarm_pass=

# Web Control
web_control_host=127.0.0.1
web_control_port=8889
web_control_static=1
web_control_auth_user=castor
web_control_auth_pass=beaver

# Monitoring Audio Stream (secured by webcontrol auth)
web_control_audio_stream=1

# Mail Fallback Notification (comment-out url to disable)
# smtp_url=smtp://mail.hello.tld:587
smtp_user=user@mail.hello.tld
smtp_pass=?
smtp_sender_name=Notifier Name
smtp_sender_address=notifier@radio.tld
smtp_recipients=tech@radio.tld, manager@radio.tld

src/Config.hpp
/*
 *  Copyright (C) 2024-2025 Christoph Pastl
 *
 *  This file is part of Castor.
 *
 *  Castor is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Lesser General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  Castor is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 *  GNU Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public License
 *  along with this program. If not, see <https://www.gnu.org/licenses/>.
 *
 *  If you use this program over a network, you must also offer access
 *  to the source code under the terms of the GNU Lesser General Public License.
 */

#pragma once

#include <fstream>
#include <sstream>
#include <string>
#include <unordered_map>
#include "util/Log.hpp"
#include "util/util.hpp"

namespace castor {
class Config {

    typedef std::unordered_map<std::string, std::string> Map;
    
    static constexpr const char* kLogLevel = "1";
    static constexpr const char* kLogPath = "./logs/castor.log";
    static constexpr const char* kSocketPath = "/tmp/castor.socket";
    static constexpr const char* kAudioSourcePath = "";
    static constexpr const char* kAudioPlaylistPath = "";
    static constexpr const char* kAudioFallbackPath = "";
    static constexpr const char* kAudioRecordPath = "";
    static constexpr const char* kRecordSchedulePath = "";
    static constexpr const char* kRecordScheduleFormat = "mp3";
    static constexpr const char* kRecordScheduleBitRate = "320000";
    static constexpr const char* kRecordBlockPath = "";
    static constexpr const char* kRecordBlockFormat = "ogg";
    static constexpr const char* kRecordBlockBitRate = "128000";
    static constexpr const char* kRecordBlockDuration = "3600";
    static constexpr const char* kDeviceName = "default";
    static constexpr const char* kStreamOutURL = "";
    static constexpr const char* kStreamOutMetadataURL = "";
    static constexpr const char* kStreamOutName = "";
    static constexpr const char* kStreamOutDescription = "";
    static constexpr const char* kStreamOutGenre = "";
    static constexpr const char* kStreamOutHREF = "";
    static constexpr const char* kProgramURL = "";
    static constexpr const char* kMediaURL = "";
    static constexpr const char* kPlaylogURL = "";
    static constexpr const char* kHealthURL = "";
    static constexpr const char* kClockURL = "";
    static constexpr const char* kCalendarRefreshInterval = "60";
    static constexpr const char* kCalendarCachePath = "./cache/calendar.json";
    static constexpr const char* kHealthReportInterval = "60";
    static constexpr const char* kYARMHost = "";
    static constexpr const char* kYARMUser = "";
    static constexpr const char* kYARMPass = "";
    static constexpr const char* kSilenceThreshold = "-80";
    static constexpr const char* kSilenceStartDuration = "5";
    static constexpr const char* kSilenceStopDuration = "1";
    static constexpr const char* kPreloadTimeFile = "3600";
    static constexpr const char* kPreloadTimeFallback = "3600";
    static constexpr const char* kProgramFadeInTime = "1.0";
    static constexpr const char* kProgramFadeOutTime = "1.0";
    static constexpr const char* kFallbackCrossFadeTime = "5.0";
    static constexpr const char* kSampleRate = "44100";
    static constexpr const char* kFallbackShuffle = "0";
    static constexpr const char* kFallbackSineSynth = "1";
    static constexpr const char* kWebControlHost = "127.0.0.1";
    static constexpr const char* kWebControlPort = "8889";
    static constexpr const char* kWebControlAuthUser = "castor";
    static constexpr const char* kWebControlAuthPass = "beaver";
    static constexpr const char* kWebControlAuthToken = "";
    static constexpr const char* kWebControlStatic = "0";
    static constexpr const char* kWebControlAudioStream = "0";
    static constexpr const char* kSMTPURL = "";
    static constexpr const char* kSMTPUser = "";
    static constexpr const char* kSMTPPass = "";
    static constexpr const char* kSMTPSenderName = "";
    static constexpr const char* kSMTPSenderAddress = "";
    static constexpr const char* kSMTPRecipients = "";

    static Map parseConfigFile(const std::string& tPath) {
        std::ifstream cfgfile(tPath);
        if (!cfgfile) {
            throw std::runtime_error("Failed to open config file");
        }
        Map cfgmap;
        std::string line;
        while (std::getline(cfgfile, line)) {
            std::istringstream iss(line);
            std::string key, value;
            if (std::getline(iss, key, '=') && std::getline(iss, value)) {
                cfgmap[key] = value;
            }
        }
        cfgfile.close();
        return cfgmap;
    }

public:

    std::string logPath;
    std::string socketPath;
    std::string audioSourcePath;
    std::string audioPlaylistPath;
    std::string audioFallbackPath;
    std::string audioRecordPath;

    std::string recordSchedulePath;
    std::string recordScheduleFormat;
    int recordScheduleBitRate;

    std::string recordBlockPath;
    std::string recordBlockFormat;
    int recordBlockBitRate;
    int recordBlockDuration;

    int preloadTimeFile;
    int preloadTimeFallback;
    int preloadTimeStream = 10;
    int preloadTimeLine = 5;

    bool fallbackShuffle;
    bool fallbackSineSynth;

    std::string iDevName;
    std::string oDevName;
    std::string streamOutURL;
    std::string streamOutMetadataURL;
    std::string streamOutName;
    std::string streamOutDescription;
    std::string streamOutGenre;
    std::string streamOutHREF;
    std::string programURL;
    std::string mediaURL;
    std::string playlogURL;
    std::string healthURL;
    std::string clockURL;
    std::string calendarCachePath;
    std::string yarmHost;
    std::string yarmUser;
    std::string yarmPass;
    std::string smtpURL;
    std::string smtpUser;
    std::string smtpPass;
    std::string smtpSenderName;
    std::string smtpSenderAddress;
    std::string smtpRecipients;
    std::string webControlHost;
    std::string webControlAuthUser;
    std::string webControlAuthPass;
    std::string webControlAuthToken;
    bool webControlStatic;
    bool webControlAudioStream;
    int webControlPort;
    int logLevel;
    int calendarRefreshInterval;
    int healthReportInterval;
    int silenceThreshold;
    int silenceStartDuration;
    int silenceStopDuration;
    int sampleRate;
    size_t samplesPerFrame = 1024;
    int streamOutBitRate = 320000;
    float outputGain = -9;
    float programFadeInTime;
    float programFadeOutTime;
    float fallbackCrossFadeTime;
    bool realtimeRendering = true;

    std::string parametersPath = "./parameters.json";

    static std::string get(Map& map, std::string mapKey, std::string defaultValue) {
        auto envKey = mapKey;
        std::transform(envKey.begin(), envKey.end(), envKey.begin(), [](unsigned char c){ return std::toupper(c); });
        std::string value;
        if (map.find(mapKey) != map.end()) value = map.at(mapKey);
        auto envVal = util::getEnvar(envKey);
        if (!envVal.empty()) value = envVal;
        if (value.empty()) value = defaultValue;
        return value;
    }

    Config(const std::string& tPath) {
        Map map = {};
        try {
            map = parseConfigFile(tPath);
        }
        catch (const std::exception& e) {
            log.error() << "Config failed to parse file: " << e.what();
        }

        logLevel = std::stoi(get(map, "log_level", kLogLevel));
        logPath = get(map, "log_path", kLogPath);
        socketPath = get(map, "socket_path", kSocketPath);
        audioSourcePath = get(map, "audio_source_path", kAudioSourcePath);
        audioPlaylistPath = get(map, "audio_playlist_path", kAudioPlaylistPath);
        audioRecordPath = get(map, "audio_record_path", kAudioRecordPath);
        audioFallbackPath = get(map, "audio_fallback_path", kAudioFallbackPath);
        iDevName = get(map, "in_device_name", kDeviceName);
        oDevName = get(map, "out_device_name", kDeviceName);
        sampleRate = std::stoi(get(map, "sample_rate", kSampleRate));
        recordSchedulePath = get(map, "record_schedule_path", kRecordSchedulePath);
        recordScheduleFormat = get(map, "record_schedule_format", kRecordScheduleFormat);
        recordScheduleBitRate = std::stoi(get(map, "record_schedule_bitrate", kRecordScheduleBitRate));
        recordBlockPath = get(map, "record_block_path", kRecordBlockPath);
        recordBlockFormat = get(map, "record_block_format", kRecordBlockFormat);
        recordBlockBitRate = std::stoi(get(map, "record_block_bitrate", kRecordBlockBitRate));
        recordBlockDuration = std::stoi(get(map, "record_block_duration", kRecordBlockDuration));
        streamOutURL = get(map, "stream_out_url", kStreamOutURL);
        streamOutMetadataURL = get(map, "stream_out_metadata_url", kStreamOutMetadataURL);
        streamOutName = get(map, "stream_out_name", kStreamOutName);
        streamOutDescription = get(map, "stream_out_description", kStreamOutDescription);
        streamOutGenre = get(map, "stream_out_genre", kStreamOutGenre);
        streamOutHREF = get(map, "stream_out_href", kStreamOutHREF);
        programURL = get(map, "program_url", kProgramURL);
        mediaURL = get(map, "media_url", kMediaURL);
        playlogURL = get(map, "playlog_url", kPlaylogURL);
        healthURL = get(map, "health_url", kHealthURL);
        clockURL = get(map, "clock_url", kClockURL);
        calendarCachePath = get(map, "calendar_cache_path", kCalendarCachePath);
        yarmHost = get(map, "yarm_host", kYARMHost);
        yarmUser = get(map, "yarm_user", kYARMUser);
        yarmPass = get(map, "yarm_pass", kYARMPass);
        calendarRefreshInterval = std::stoi(get(map, "calendar_refresh_interval", kCalendarRefreshInterval));
        healthReportInterval = std::stoi(get(map, "health_report_interval", kHealthReportInterval));
        silenceThreshold = std::stoi(get(map, "silence_threshold", kSilenceThreshold));
        silenceStartDuration = std::stoi(get(map, "silence_start_duration", kSilenceStartDuration));
        silenceStopDuration = std::stoi(get(map, "silence_stop_duration", kSilenceStopDuration));
        preloadTimeFile = std::stoi(get(map, "preload_time_file", kPreloadTimeFile));
        preloadTimeFallback = std::stoi(get(map, "preload_time_fallback", kPreloadTimeFallback));
        programFadeInTime = std::stof(get(map, "program_fade_in_time", kProgramFadeInTime));
        programFadeOutTime = std::stof(get(map, "program_fade_out_time", kProgramFadeOutTime));
        fallbackCrossFadeTime = std::stof(get(map, "fallback_cross_fade_time", kFallbackCrossFadeTime));
        fallbackShuffle = std::stoi(get(map, "fallback_shuffle", kFallbackShuffle));
        fallbackSineSynth = std::stoi(get(map, "fallback_sine_synth", kFallbackSineSynth));
        smtpURL = get(map, "smtp_url", kSMTPURL);
        smtpUser = get(map, "smtp_user", kSMTPUser);
        smtpPass = get(map, "smtp_pass", kSMTPPass);
        smtpSenderName = get(map, "smtp_sender_name", kSMTPSenderName);
        smtpSenderAddress = get(map, "smtp_sender_address", kSMTPSenderAddress);
        smtpRecipients = get(map, "smtp_recipients", kSMTPRecipients);
        webControlHost = get(map, "web_control_host", kWebControlHost);
        webControlPort = std::stoi(get(map, "web_control_port", kWebControlPort));
        webControlAuthUser = get(map, "web_control_auth_user", kWebControlAuthUser);
        webControlAuthPass = get(map, "web_control_auth_pass", kWebControlAuthPass);
        webControlAuthToken = get(map, "web_control_auth_token", kWebControlAuthToken);
        webControlStatic = std::stoi(get(map, "web_control_static", kWebControlStatic));
        webControlAudioStream = std::stoi(get(map, "web_control_audio_stream", kWebControlAudioStream));
        
        log.info() << "Config:"
        << "\n\t logLevel=" << logLevel
        << "\n\t logPath=" << logPath
        << "\n\t socketPath=" << socketPath
        << "\n\t audioSourcePath=" << audioSourcePath
        << "\n\t audioPlaylistPath=" << audioPlaylistPath
        << "\n\t audioFallbackPath=" << audioFallbackPath
        << "\n\t audioRecordPath=" << audioRecordPath
        << "\n\t recordSchedulePath=" << recordSchedulePath
        << "\n\t recordScheduleFormat=" << recordScheduleFormat
        << "\n\t recordScheduleBitRate=" << recordScheduleBitRate
        << "\n\t recordBlockPath=" << recordBlockPath
        << "\n\t recordBlockFormat=" << recordBlockFormat
        << "\n\t recordBlockBitRate=" << recordBlockBitRate
        << "\n\t recordBlockDuration=" << recordBlockDuration
        << "\n\t iDevName=" << iDevName
        << "\n\t oDevName=" << oDevName
        << "\n\t sampleRate=" << sampleRate
        << "\n\t streamOutURL=" << streamOutURL
        << "\n\t streamOutMetadataURL=" << streamOutMetadataURL
        << "\n\t streamOutName=" << streamOutName
        << "\n\t streamOutDescription=" << streamOutDescription
        << "\n\t streamOutGenre=" << streamOutGenre
        << "\n\t streamOutHREF=" << streamOutHREF
        << "\n\t programURL=" << programURL
        << "\n\t mediaURL=" << mediaURL
        << "\n\t playlogURL=" << playlogURL
        << "\n\t healthURL=" << healthURL
        << "\n\t clockURL=" << clockURL
        << "\n\t calendarRefreshInterval=" << calendarRefreshInterval
        << "\n\t healthReportInterval=" << healthReportInterval
        << "\n\t calendarCachePath=" << calendarCachePath
        << "\n\t yarmHost=" << yarmHost
        << "\n\t yarmUser=" << yarmUser
        << "\n\t smtpURL=" << smtpURL
        << "\n\t smtpUser=" << smtpUser
        << "\n\t smtpSenderName=" << smtpSenderName
        << "\n\t smtpSenderAddress=" << smtpSenderAddress
        << "\n\t smtpRecipients=" << smtpRecipients
        << "\n\t webControlHost=" << webControlHost
        << "\n\t webControlPort=" << webControlPort
        << "\n\t webControlStatic=" << webControlStatic
        << "\n\t webControlAudioStream=" << webControlAudioStream
        << "\n\t webControlAuthUser=" << webControlAuthUser
        << "\n\t silenceThreshold=" << silenceThreshold
        << "\n\t silenceStartDuration=" << silenceStartDuration
        << "\n\t silenceStopDuration=" << silenceStopDuration
        << "\n\t preloadTimeFile=" << preloadTimeFile
        << "\n\t preloadTimeFallback=" << preloadTimeFallback
        << "\n\t programFadeInTime=" << programFadeInTime
        << "\n\t programFadeOutTime=" << programFadeOutTime
        << "\n\t fallbackCrossFadeTime=" << fallbackCrossFadeTime
        << "\n\t fallbackSineSynth=" << fallbackSineSynth
        << "\n\t fallbackShuffle=" << fallbackShuffle;
    }
};
}